File: app/src/main/java/com/hundaol/ethiopic/stamps/MonthStamp.java
package com.hundaol.ethiopic.stamps;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;

import com.hundaol.ethiocal.R;
import com.hundaol.ethiopic.cal.ICal;
import com.hundaol.ethiopic.views.CalendarViewModel;

/**
 * Created by john.pirie on 2017-05-01.
 */

public class MonthStamp {
    private final Context context;
    private CalendarViewModel viewModel;
    private final DayStamp dayStamp;
    private final LabelStamp labelStamp;

    public int jdn;

    public final Paint forePaint;

    public MonthStamp(Context context, CalendarViewModel viewModel) {
        this.context = context;
        this.viewModel = viewModel;

        dayStamp = new DayStamp(context, viewModel);
        labelStamp = new LabelStamp(context, viewModel);

        forePaint = new Paint();
        forePaint.setStyle(Paint.Style.STROKE);
        forePaint.setColor(context.getResources().getColor(R.color.month_foreground));
        forePaint.setStrokeWidth(context.getResources().getDimensionPixelSize(R.dimen.month_border_width));
    }

    public void setJdn(int jdn) {
        this.jdn = jdn;
    }

    public void stamp(Canvas canvas) {
        int jdn = viewModel.getCal().firstOfMonth(this.jdn);

        ICal cal = viewModel.getCal();
        for (int i = 0; i < cal.getDaysInMonth(jdn); i++) {
            dayStamp.setJdn(jdn + i);
            dayStamp.stamp(canvas);
        }

        labelStamp.setJdn(jdn);
        labelStamp.stamp(canvas);

        Path monthPath = viewModel.pathForMonth(jdn);
        canvas.drawPath(monthPath, forePaint);
    }
}

File: app/src/main/java/com/hundaol/ethiopic/stamps/MonthStamp.kt
package com.hundaol.ethiopic.stamps

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import com.hundaol.ethiopic.cal.GregorianCal
import com.hundaol.ethiopic.cal.ICal
import com.hundaol.ethiopic.domain.ColorModel
import com.hundaol.ethiopic.domain.DateModel
import com.hundaol.ethiopic.views.calendar.CalendarViewModel

/**
 * Created by john.pirie on 2017-06-06.
 */
class MonthStamp(var jdn: Int = 0) {

    var viewModel = CalendarViewModel()
        get() = field
        set(value) {
            field = value
            dayStamp.viewModel = viewModel
            labelStamp.viewModel = viewModel
        }

    var dateModel = DateModel.default
        get() = field
        set(value) {
            field = value
            dayStamp.dateModel = dateModel
            labelStamp.dateModel = dateModel
        }

    var colorModel = ColorModel.default
        get() = field
        set(value) {
            field = value
            dayStamp.colorModel = colorModel
            labelStamp.colorModel = colorModel
        }

    var cal: ICal = GregorianCal.INSTANCE
        get() = field
        set(value) {
            field = value
            dayStamp.cal = value
            labelStamp.cal = value
        }

    val bounds: RectF
        get() = viewModel.boundsForMonth(dateModel, cal, jdn)

    val backPaint = Paint()
    val dayStamp = DayStamp()
    val labelStamp = LabelStamp()

    init {
        backPaint.style = Paint.Style.FILL_AND_STROKE
    }

    fun stamp(canvas: Canvas) {
        backPaint.color = colorModel.backgroundColorForMonth(cal, jdn)
        canvas.drawPath(viewModel.pathForMonth(dateModel, cal, jdn), backPaint)

        var j = cal.firstOfMonth(jdn)

        for (i in 0..cal.getDaysInMonth(j) - 1) {
            dayStamp.jdn = j + i
            dayStamp.stamp(canvas)
        }

        labelStamp.jdn = jdn
        labelStamp.stamp(canvas)

    }
}